### PR TITLE
New classifiers

### PR DESCRIPTION
**Pull request by teaearlgraycold**
## What does this PR do?

Adds 7 new classifiers to TPOT

## Where should the reviewer start?

At the _ada_boost() method in the TPOT class, going all the way down to the _p_aggr() method.

There is also relevant export code and docs that should be checked for an LGTM.

## How should this PR be tested?

Travis should test the classifiers themselves, but a few pipelines could be made and exported to confirm that the code indeed works.

## What are the relevant issues?

#128

## Questions:

- Do the docs need to be updated?

Yes. I have already updated them.

- Does this PR add new (Python) dependencies?

No.

### CODE REVIEW

File: tpot/tpot.py [hunk 1]
```diff
@@ -496,6 +500,181 @@ def _random_forest(self, input_df, max_features):
...
+            Learning rate shrinks the contribution of each classifier by learning_rate.
+
+        Returns
+        -------
+        input_df: pandas.DataFrame {n_samples, n_features+['guess', 'group', 'class', 'SyntheticFeature']}
+            Returns a modified input DataFrame with the guess column updated according to the classifier's predictions.
+            Also adds the classifiers's predictions as a 'SyntheticFeature' column.
+
+        """
+        return self._train_model_and_predict(input_df, AdaBoostClassifier,
```
Comment (rhiever) [inline tpot/tpot.py]: `learning_rate` should be capped at > 0.

Looking at the docs further, I also think that for the AdaBoostClassifier, we should allow `n_estimators` to be evolved as well, with a max of 500 estimators. The AdaBoostClassifier is one unique case where there is a tradeoff between `n_estimators` and `learning_rate`.

File: tpot/tpot.py [hunk 2]
```diff
@@ -496,6 +500,181 @@ def _random_forest(self, input_df, max_features):
...
+            Returns a modified input DataFrame with the guess column updated according to the classifier's predictions.
+            Also adds the classifiers's predictions as a 'SyntheticFeature' column.
+
+        """
+        # Select criterion string from list of valid parameters
+        criterion_values = ['gini', 'entropy']
+        criterion_selection = criterion_values[criterion % len(criterion_values)]
+
+        return self._train_model_and_predict(input_df, ExtraTreesClassifier,
+            criterion=criterion_selection, max_features=max_features,
```
Comment (rhiever) [inline tpot/tpot.py]: `max_features` should be capped between 1 and `len(num_training_features)`.

File: tpot/tpot.py [hunk 3]
```diff
@@ -496,6 +500,181 @@ def _random_forest(self, input_df, max_features):
...
+            Returns a modified input DataFrame with the guess column updated according to the classifier's predictions.
+            Also adds the classifiers's predictions as a 'SyntheticFeature' column.
+
+        """
+        fit_bool = (fit_intercept % 2) == 0
+
+        loss_values = ['hinge', 'squared_hinge']
+        loss_selection = loss_values[loss % len(loss_values)]
+
+        return self._train_model_and_predict(input_df, LinearSVC, C=C,
```
Comment (rhiever) [inline tpot/tpot.py]: `C` should be capped at >0.

File: tpot/tpot.py (hunk 1 above)
Comment (teaearlgraycold) [inline tpot/tpot.py]: Should I just do

```Python
max(learning_rate, 0.001)
```

Not sure what the exact minimum should be here.

File: tpot/tpot.py [hunk 4]
```diff
@@ -496,6 +500,181 @@ def _random_forest(self, input_df, max_features):
...
+            loss=loss_selection, fit_intercept=fit_bool, random_state=42)
+
+    def _p_aggr(self, input_df, C, loss, fit_intercept):
```
Comment (rhiever) [inline tpot/tpot.py]: Please make this `passive_aggressive()`. `p_aggr` is unclear.

File: tpot/tpot.py [hunk 5]
```diff
@@ -496,6 +500,181 @@ def _random_forest(self, input_df, max_features):
...
+            Returns a modified input DataFrame with the guess column updated according to the classifier's predictions.
+            Also adds the classifiers's predictions as a 'SyntheticFeature' column.
+
+        """
+        fit_bool = (fit_intercept % 2) == 0
+
+        loss_values = ['hinge', 'squared_hinge']
+        loss_selection = loss_values[loss % len(loss_values)]
+
+        return self._train_model_and_predict(input_df, PassiveAggressiveClassifier,
```
Comment (rhiever) [inline tpot/tpot.py]: `C` should be capped at >0.

**Comment by rhiever** [comment]
When writing test cases for classifiers, test with normal parameters as well as extreme parameters: negative values, out-of-bounds values, etc. That will help catch issues where we're allowing invalid parameters to be passed to the various models.

File: tpot/tpot.py (hunk 1 above)
Comment (rhiever) [inline tpot/tpot.py]: Check `_logistic_regression` for an example: https://github.com/rhiever/tpot/blob/master/tpot/tpot.py#L516

```python
C = max(0.0001, C)
```

0.0001 seems like a fine minimum value for now.

**Comment by teaearlgraycold** [comment]
Now addresses #151

**Pull request merged**